### DISCUSSION TITLE
✨ Access Service stdout/stderr and exit code

### DISCUSSION THREAD

**Discussion by jedevc** [dagger/dagger#12449] [Ideas]
### What are you trying to do?

Currently, once a service `start`s, there's not a ton of interaction points for it. We can `stop` it... and that's about it.

One key thing I'd like is access to the stdout/stderr of a service. At the moment, these are exclusively streamed to the progrock output, but it would be nice to be able to output them somewhere else.

> [!NOTE]
> In addition to this, I'd *really* like to have a way to suppress the default progrock service output. These in combination would be *so* useful for our own integration tests, we could redirect all the output from the engine logs, and output them into their own github step, so that the main step isn't like 200k long and 50MB+ in size :smile:

We *could* add `stdout` and `stderr` fields to a service, that returns the logs up to that point. However - we'd still need to be storing the output (since we don't necessarily *know* if `stdout`/`stderr` would be called later), and if the service is chatty, this could use a lot of unnecessary space. We could allow pre-configuration when running the service?

The *ideal* API to me would be to get a `stdout` *stream* - and then dagger would stream the result back to me (this would also be super useful for the normal `Container.stdout` as well tbf). But this also seems kinda logistically tricky, and feels tricky to integrate to SDKs.

---

Secondly, I'd also *really* like access to the status code when we call `stop`. This one feels a little bit easier to actually do the plumbing, but it's still quite unclear as to where this field should live. Just having a `Service.exitCode` field works, but it feels weird to be able to call that on a service that's never even started.

As an example, I'd like to refactor `TestEngineExitsZeroOnSignal` to use the new `Service.stop` SIGTERM behavior:

```go
func TestEngineExitsZeroOnSignal(t *testing.T) {
	t.Parallel()
	c, ctx := connect(t)

	devEngine := devEngineContainer(c).
		WithMountedCache("/var/lib/dagger", c.CacheVolume("dagger-dev-engine-state-"+identity.NewID())).
		WithExec([]string{"--addr", "tcp://0.0.0.0:1234"}, dagger.ContainerWithExecOpts{
			InsecureRootCapabilities: true,
		}).
		AsService()

	// engine starts, and we only return after the healthcheck starts passing
	devEngine, err := devEngine.Start(ctx)
	require.NoError(t, err)

	// engine should shutdown with exit code 0 when receiving SIGTERM
	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()

	// NOTE: there's no way to check the exit code!
	_, err = devEngine.Stop(ctx)
	require.NoError(t, err)
}
```

One idea I had was to somehow include the status code in an error - but should `Stop` return an error on a non-zero exit code?

**Comment by gerhard** [comment]
I like both ideas very much!

I recall that @helderco has been thinking & working on this in the context of regular operations, and propagating error codes correctly. Even if I misremember, what are your thoughts @helder about this issue?

In terms of returning an exit code for a service, I really think that this should be typed, possibly drawing inspiration from https://pkg.go.dev/github.com/sean-/sysexits . If we did that, we could introduce a special exit code when the service is still running, and then the `String()` value for that exit code could be `""`

If we did the above, then the `Service` type could add a new `ExitCode()` function. This wouldn't change the `Service.Stop()` function API, which means a nice additive change for end users.

Similar to `Service.ExitCode()`, what are your thoughts on adding `Stdout()` & `Stderr()` to the `Service` type? I know that the `Container` type already has this, but I don't think that using these from a `Service` type is currently possible. Would that solve (at least partly) your issue @jedevc? WDYT @vito?

cc @shykes @sipsma @aluzzardi

**Comment by bedilbek** [comment]
What is the update on this issue? Is there any progress?

**Comment by bedilbek** [comment]
Is there any progress on this?

**Comment by bedilbek** [comment]
@jedevc
